### ISSUE TITLE
How to Handle Storage Issue in Windows IoT core running Azure IoT edge services. 

### ISSUE THREAD

**Issue by arnabhetc** [Azure/iotedge#1598]
Hi,
 
We using IoT edge services to develop a production-level solution, where more than 100 gateways will be deployed, We are using Aeeon Gateways along with windows IoT core to run IoT edge services. 

The problem is - 

-  we are not able to handle the storage of the gateway, and it's getting full after somedays,  the gateways are running only IoT-edge services nothing else. 

- we need to run the gateways for months without any problem in the production environment, but today after 24 days of running Edge services both the gateway were crashed 

![Test-gateway](https://user-images.githubusercontent.com/47913360/63355730-3ee7f880-c384-11e9-8920-be992ee21fe8.jpg)
This gateway was showing the error code - **0xC000021A**

![Dev gatway](https://user-images.githubusercontent.com/47913360/63355780-558e4f80-c384-11e9-9b4e-c643512df99b.gif)

This gateway shows nothing apart from the this emoji 

We are using the offline capabilities fo the IoT-edge, bellow are settings 
   
    "storeAndForwardConfiguration": {
          "timeToLiveSecs": 86400
        }

My questions

1.  what is causing the storage issue in the gateway?
2.  How to **safely** clear all the stored data related to IoT-edge in the gateway, so whenever storage is full or about to full we can delete those files. 
3. How to handle storage for custom modules written by us
4. we have tried to remove docker-image from the IoT-core device by running bellow command, but it failed, The command didn't work for Windows IoT core, we tired it to remove the image from the device so all data associated to it get deleted but **it didn't work**.

     docker rmi $(docker images -q)

  @iotedge-ci  @azure-sdk

**Comment by veyalla** [comment]
Please share the output of 'iotedge check'.

**Comment by arnabhetc** [comment]
[192.168.1.104]: PS C:\Data\Users\administrator.MINWINPC\Documents> iotedge check
Configuration checks
--------------------
√ config.yaml is well-formed - OK
× config.yaml has well-formed connection string - Error
    Device is not using manual provisioning, so Azure IoT Hub hostname needs to be specified with --iothub-hostname
√ container engine is installed and functional - OK
√ Windows host version is supported - OK
√ config.yaml has correct hostname - OK
× config.yaml has correct URIs for daemon mgmt endpoint - Error
    Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.0.8' locally
    1.0.8: Pulling from azureiotedge-diagnostics
    6a5e981230da: Pulling fs layer
    f7de0eb29689: Pulling fs layer
    89988ae4dec4: Pulling fs layer
    89988ae4dec4: Verifying Checksum
    89988ae4dec4: Download complete
    6a5e981230da: Retrying in 5 seconds
    f7de0eb29689: Retrying in 5 seconds
    6a5e981230da: Retrying in 4 seconds
    f7de0eb29689: Retrying in 4 seconds
    6a5e981230da: Retrying in 3 seconds
    f7de0eb29689: Retrying in 3 seconds
    6a5e981230da: Retrying in 2 seconds
    f7de0eb29689: Retrying in 2 seconds
    6a5e981230da: Retrying in 1 second
    f7de0eb29689: Retrying in 1 second
    6a5e981230da: Retrying in 10 seconds
    f7de0eb29689: Retrying in 10 seconds
    6a5e981230da: Retrying in 9 seconds
    f7de0eb29689: Retrying in 9 seconds
    6a5e981230da: Retrying in 8 seconds
    f7de0eb29689: Retrying in 8 seconds
    6a5e981230da: Retrying in 7 seconds
    f7de0eb29689: Retrying in 7 seconds
    6a5e981230da: Retrying in 6 seconds
    f7de0eb29689: Retrying in 6 seconds
    6a5e981230da: Retrying in 5 seconds
    f7de0eb29689: Retrying in 5 seconds
    6a5e981230da: Retrying in 4 seconds
    f7de0eb29689: Retrying in 4 seconds
    6a5e981230da: Retrying in 3 seconds
    f7de0eb29689: Retrying in 3 seconds
    6a5e981230da: Retrying in 2 seconds
    f7de0eb29689: Retrying in 2 seconds
    6a5e981230da: Retrying in 1 second
    f7de0eb29689: Retrying in 1 second
    6a5e981230da: Retrying in 15 seconds
    f7de0eb29689: Retrying in 15 seconds
    6a5e981230da: Retrying in 14 seconds
    f7de0eb29689: Retrying in 14 seconds
    6a5e981230da: Retrying in 13 seconds
    f7de0eb29689: Retrying in 13 seconds
    6a5e981230da: Retrying in 12 seconds
    f7de0eb29689: Retrying in 12 seconds
    f7de0eb29689: Retrying in 11 seconds
    6a5e981230da: Retrying in 11 seconds
    6a5e981230da: Retrying in 10 seconds
    f7de0eb29689: Retrying in 10 seconds
    f7de0eb29689: Retrying in 9 seconds
    6a5e981230da: Retrying in 9 seconds
    f7de0eb29689: Retrying in 8 seconds
    6a5e981230da: Retrying in 8 seconds
    f7de0eb29689: Retrying in 7 seconds
    6a5e981230da: Retrying in 7 seconds
    6a5e981230da: Retrying in 6 seconds
    f7de0eb29689: Retrying in 6 seconds
    f7de0eb29689: Retrying in 5 seconds
    6a5e981230da: Retrying in 5 seconds
    f7de0eb29689: Retrying in 4 seconds
    6a5e981230da: Retrying in 4 seconds
    6a5e981230da: Retrying in 3 seconds
    f7de0eb29689: Retrying in 3 seconds
    6a5e981230da: Retrying in 2 seconds
    f7de0eb29689: Retrying in 2 seconds
    f7de0eb29689: Retrying in 1 second
    6a5e981230da: Retrying in 1 second
    6a5e981230da: Retrying in 20 seconds
    f7de0eb29689: Retrying in 20 seconds
    6a5e981230da: Retrying in 19 seconds
    f7de0eb29689: Retrying in 19 seconds
    6a5e981230da: Retrying in 18 seconds
    f7de0eb29689: Retrying in 18 seconds
    6a5e981230da: Retrying in 17 seconds
    f7de0eb29689: Retrying in 17 seconds
    6a5e981230da: Retrying in 16 seconds
    f7de0eb29689: Retrying in 16 seconds
    6a5e981230da: Retrying in 15 seconds
    f7de0eb29689: Retrying in 15 seconds
    6a5e981230da: Retrying in 14 seconds
    f7de0eb29689: Retrying in 14 seconds
    6a5e981230da: Retrying in 13 seconds
    f7de0eb29689: Retrying in 13 seconds
    6a5e981230da: Retrying in 12 seconds
    f7de0eb29689: Retrying in 12 seconds
    6a5e981230da: Retrying in 11 seconds
    f7de0eb29689: Retrying in 11 seconds
    6a5e981230da: Retrying in 10 seconds
    f7de0eb29689: Retrying in 10 seconds
    6a5e981230da: Retrying in 9 seconds
    f7de0eb29689: Retrying in 9 seconds
    f7de0eb29689: Retrying in 8 seconds
    6a5e981230da: Retrying in 8 seconds
    6a5e981230da: Retrying in 7 seconds
    f7de0eb29689: Retrying in 7 seconds
    6a5e981230da: Retrying in 6 seconds
    f7de0eb29689: Retrying in 6 seconds
    6a5e981230da: Retrying in 5 seconds
    f7de0eb29689: Retrying in 5 seconds
    6a5e981230da: Retrying in 4 seconds
    f7de0eb29689: Retrying in 4 seconds
    6a5e981230da: Retrying in 3 seconds
    f7de0eb29689: Retrying in 3 seconds
    6a5e981230da: Retrying in 2 seconds
    f7de0eb29689: Retrying in 2 seconds
    f7de0eb29689: Retrying in 1 second
    6a5e981230da: Retrying in 1 second
    docker: read tcp 192.168.1.104:50767->204.79.197.219:443: wsarecv: An existing connection was forcibly closed by the remote host.
    See 'docker run --help'.
√ latest security daemon - OK
√ host time is close to real time - OK
× container time is close to host time - Error
    Could not query local time inside container
‼ DNS server - Warning
    Container engine is not configured with DNS server setting, which may impact connectivity to IoT Hub.
    Please see https://aka.ms/iotedge-prod-checklist-dns for best practices.
    You can ignore this warning if you are setting DNS server per module in the Edge deployment.
‼ production readiness: certificates - Warning
    Device is using self-signed, automatically generated certs.
    Please see https://aka.ms/iotedge-prod-checklist-certs for best practices.
√ production readiness: certificates expiry - OK
√ production readiness: container engine - OK
‼ production readiness: logs policy - Warning
    Container engine is not configured to rotate module logs which may cause it run out of disk space.
    Please see https://aka.ms/iotedge-prod-checklist-logs for best practices.
    You can ignore this warning if you are setting log policy per module in the Edge deployment.

Connectivity checks
-------------------
√ host can connect to and perform TLS handshake with DPS endpoint - OK
√ Edge Hub can bind to ports on host - OK

10 check(s) succeeded.
3 check(s) raised warnings. Re-run with --verbose for more details.
3 check(s) raised errors. Re-run with --verbose for more details.
6 check(s) were skipped due to errors from other checks. Re-run with --verbose for more details.
iotedge :
    + CategoryInfo          : NotSpecified: (:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

**Comment by arnabhetc** [comment]
Hi,
I have tried the following command 

    [192.168.1.104]: PS C:\Data\Users\administrator.MINWINPC\Documents> docker image prune -a
    WARNING! This will remove all images without at least one container associated with them.
    Are you sure you want to continue? [y/N] Total reclaimed space: 0B

The problem is here we don't have any untagged containers,  that's why it didn't serve the purpose,

Here we are running container in both the Linux and Windows IoT core environment, and in both cases, the storage of the device is getting full because of running containers,  we want a fix for both the cases.

**Comment by veyalla** [comment]
How have you determined that the problem is due to running out of disk space?

If you suspect that, see recommendations from iotedge check: 

‼ production readiness: logs policy - Warning
Container engine is not configured to rotate module logs which may cause it run out of disk space.
Please see https://aka.ms/iotedge-prod-checklist-logs for best practices.
You can ignore this warning if you are setting log policy per module in the Edge deployment.

**Comment by lucianoctorres** [comment]
May be related or not, but since the last update of iotedge I am running out of space on my Ubuntu boxes as well. My docker partition was getting full, so I moved the Edge Hub storage out of the container -- it has more space so it doesn't crash docker anymore, but the storage grows steadily, I have now 10G of storage for edge hub after two days running.

The logs are all rotating (per https://docs.microsoft.com/en-us/azure/iot-edge/production-checklist#place-limits-on-log-size). If I restart edge hub (just with iotedge restart edgeHub) the storage shrinks to a few megs.

**Comment by veyalla** [comment]
@lucianoctorres It might depend on the ingestion rate through edgeHub and the [message TTL settings](https://docs.microsoft.com/en-us/azure/iot-edge/offline-capabilities#time-to-live).

**Comment by lucianoctorres** [comment]
@veyalla TTL is the default 7200 secs. The rate of ingestion hasn’t changed much in the last 7 months, but only recently I started having this problem

**Comment by ancaantochi** [comment]
@lucianoctorres, 
Can you please add logs to for edgeHub, edgeAgent and iotedged? 
After you restart edgeHub does the storage shrinks immediately or it takes some time? Which version of edgeHub are you using?

Thanks,
Anca

**Comment by lucianoctorres** [comment]
Hi Anca

the versions are 

```
mcr.microsoft.com/azureiotedge-hub                           1.0                              3d5d0c5ccf62        9 days ago          150MB
mcr.microsoft.com/azureiotedge-agent                         1.0                              e1a90ff7d694        9 days ago          137MB
```
and
```
iotedge --version
iotedge 1.0.8 (208b2204fd30e856d00b280112422130c104b9f0)
```

When I restart iotedge, the storage remains until it's able to establish the proper connection with Azure IoT Hub, and when it happens the storage shrinks immediately.

I am going to get the logs for you, but there is nothing special about them. For edgeHub, the pattern is

``
<7> 2019-09-06 14:13:05.039 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint <hidden>/input with batch size 0. Next start offset = 909729.
<7> 2019-09-06 14:13:05.099 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint <hidden>/input starting from 0 with batch size 10.
<7> 2019-09-06 14:13:05.099 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint <hidden>/input starting from 0 with batch size 10.
<7> 2019-09-06 14:13:05.099 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint <hidden>/input with batch size 0. Next start offset = 0.
<7> 2019-09-06 14:13:05.099 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint <hidden>/input with batch size 0. Next start offset = 0.
<7> 2019-09-06 14:13:05.103 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Getting next batch for endpoint iothub starting from 1583009 with batch size 100.
<7> 2019-09-06 14:13:05.103 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Storage.MessageStore] - Obtained next batch for endpoint iothub with batch size 0. Next start offset = 1583009.
``

**Comment by ancaantochi** [comment]
Thanks! I was able to reproduce and I am investigating the root cause, will update here when I have more details.

**Comment by lt72** [comment]
This issue has been fixed in 1.0.8.3.
